server/src-lib/Hasura/Class.hs

```diff
@@ -0,0 +1,84 @@
+module Hasura.Class
+  ( MetadataStorageT(..)
+  , runMetadataStorageT
+  , MonadMetadataStorage(..)
+  )
+where
+
+import           Hasura.Eventing.HTTP
+import           Hasura.Eventing.ScheduledTrigger.Types
+import           Hasura.Prelude
+import           Hasura.RQL.Types
+
+import           Control.Monad.Morph                    (MFunctor, hoist)
+
+import qualified Hasura.Tracing                         as Tracing
+
+newtype MetadataStorageT m a
+  = MetadataStorageT {unMetadataStorageT :: ExceptT QErr m a}
+  deriving ( Functor, Applicative, Monad
+           , MonadError QErr
+           , MFunctor
+           , MonadTrans
+           , MonadIO
+           )
+
+runMetadataStorageT
+  :: MetadataStorageT m a -> m (Either QErr a)
+runMetadataStorageT =
+  runExceptT . unMetadataStorageT
+
+
+class (Monad m) => MonadMetadataStorage m where
+
+  -- Scheduled triggers
+  getDeprivedCronTriggerStats :: MetadataStorageT m [CronTriggerStats]
+  getScheduledEventsForDelivery :: MetadataStorageT m ([CronEvent], [OneOffScheduledEvent])
+  insertScheduledEvent :: ScheduledEventSeed -> MetadataStorageT m ()
```

Comment:
Does this also make sense?
```suggestion
class (MonadError QErr m) => MonadMetadataStorage m where

  -- Scheduled triggers
  getDeprivedCronTriggerStats :: m [CronTriggerStats]
  getScheduledEventsForDelivery :: m ([CronEvent], [OneOffScheduledEvent])
  insertScheduledEvent :: ScheduledEventSeed -> m ()
```

So rather than writing an instance for `ServerAppM`:

```haskell
instance MonadMetadataStorage ServerAppM where
```

you would write an instance for `MonadStorageT ServerAppM`:

```haskell
instance MonadMetadataStorage (MonadStorageT ServerAppM) where
```

Reply:
Please see my experiment [here](https://github.com/rakeshkky/graphql-engine/pull/8).

Reply:
Implemented!

Reply:
I'm resolving this particular discussion, although spiritually @lexi-lambda has struck similar points below and I think that discussion should be continued.